common/crypto/platform/openssl/crypto.cpp

```diff
@@ -66,6 +66,43 @@ auto bio_free_all_func = [](BIO *bio) {
 	}
 };
 
+// NOTE: GetOpenSSLErrorMessage should be called upon all OpenSSL errors, as
+// the errors are queued, and if not harvested, the FIFO structure of the
+// queue will mean that if you just get one, you might actually get the wrong
+// one.
+string GetOpenSSLErrorMessage() {
+	const auto sysErrorCode = errno;
+	auto sslErrorCode = ERR_get_error();
+
+	std::string errorDescription;
+	if (sslErrorCode != 0) {
+		errorDescription = ERR_error_string(sslErrorCode, nullptr);
+		while (sslErrorCode != 0) {
+			if (!errorDescription.empty()) {
+				errorDescription += '\n';
+			}
+			errorDescription += ERR_error_string(sslErrorCode, nullptr);
+			sslErrorCode = ERR_get_error();
+		}
+		std::string errorDescription;
+		while (sslErrorCode != 0) {
+			errorDescription += ERR_error_string(sslErrorCode, nullptr);
+			sslErrorCode = ERR_get_error();
+		}
+		if (sysErrorCode != 0) {
+			if (!errorDescription.empty()) {
+				errorDescription += '\n';
+			}
+			errorDescription += "System error, code=" + std::to_string(sysErrorCode);
+			errorDescription += ", ";
+			errorDescription += strerror(sysErrorCode);
+		}
+		return errorDescription;
+	}
+	return "";
```

Comment:
Is it possible that we don't have an ssl error but we do have a system error? If so the function would return `"`.

Reply:
Good point! I guess it _could happen_. Let's just say we will never forget logging the error :p